File: building-blocks/read-data-feeds/read-data-feeds-go/my-workflow/workflow.go
package main

import (
	"fmt"
	"log/slog"

	"github.com/shopspring/decimal"

	"github.com/ethereum/go-ethereum/common"

	"my-project/contracts/evm/src/generated/btcusd_price_feed"

	"github.com/smartcontractkit/cre-sdk-go/capabilities/blockchain/evm"
	"github.com/smartcontractkit/cre-sdk-go/capabilities/scheduler/cron"
	"github.com/smartcontractkit/cre-sdk-go/cre"
)

type Config struct {
	// 6-field cron; this fires on the 0th second every 10 minutes
	Schedule          string `json:"schedule"`          // e.g. "0 */10 * * * *"
	ChainName         string `json:"chainName"`         // e.g. "ethereum-mainnet-arbitrum-1"
	AggregatorAddress string `json:"aggregatorAddress"` // e.g. "0x6ce185860a4963106506C203335A2910413708e9"
}

func InitWorkflow(cfg *Config, logger *slog.Logger, _ cre.SecretsProvider) (cre.Workflow[*Config], error) {
	return cre.Workflow[*Config]{
		cre.Handler(cron.Trigger(&cron.Config{Schedule: cfg.Schedule}), onTick),
	}, nil
}

func onTick(cfg *Config, runtime cre.Runtime, _ *cron.Payload) (string, error) {
	lg := runtime.Logger()

	// Build EVM client from chain name
	selector, err := evm.ChainSelectorFromName(cfg.ChainName)
	if err != nil {
		return "", fmt.Errorf("chain selector: %w", err)
	}
	client := &evm.Client{ChainSelector: selector}

	// Bind the feed proxy.
	addr := common.HexToAddress(cfg.AggregatorAddress)
	feed, err := btcusd_price_feed.NewBTCUSDPriceFeed(client, addr, nil)
	if err != nil {
		return "", fmt.Errorf("binding: %w", err)
	}

	// Reads default to the finalized block when you pass nil
	decimals, err := feed.Decimals(runtime, nil).Await()
	if err != nil {
		lg.Error("decimals() failed", "err", err)
		return "", err
	}

	raw, err := feed.LatestAnswer(runtime, nil).Await()
	if err != nil {
		lg.Error("latestAnswer() failed", "err", err)
		return "", err
	}

	// Scale latestAnswer by 10^decimals.
	scale := decimal.New(1, int32(decimals)) // = 10^decimals
	scaled := decimal.NewFromBigInt(raw, 0).Div(scale)

	lg.Info("BTC/USD feed read",
		"chain", cfg.ChainName,
		"address", cfg.AggregatorAddress,
		"decimals", decimals,
		"latestAnswerRaw", raw.String(),
		"latestAnswerScaled", scaled.String(),
	)

	// Return the human-readable number
	return scaled.String(), nil
}

File: building-blocks/read-data-feeds/read-data-feeds-ts/my-workflow/workflow.ts
import {
	cre,
	getNetwork,
	type Runtime,
	type CronPayload,
} from '@chainlink/cre-sdk';
import { formatUnits, type Address } from 'viem';
import { z } from 'zod';
import { PriceFeedAggregator } from '../contracts/evm/ts/generated/PriceFeedAggregator';

// ---------- Config ----------

export const configSchema = z.object({
	// e.g. "0 */10 * * * *" (every 10 minutes, at second 0)
	schedule: z.string(),
	// e.g. "ethereum-mainnet-arbitrum-1"
	chainName: z.string(),
	// list of feeds (BTC/USD, ETH/USD, ...)
	feeds: z.array(
		z.object({
			name: z.string(),    // "BTC/USD"
			address: z.string(), // proxy address
		}),
	),
});

type Config = z.infer<typeof configSchema>;

type PriceResult = {
	name: string;
	address: string;
	decimals: number;
	latestAnswerRaw: string;
	scaled: string;
};

// ---------- Helpers ----------

function getEvmClient(chainName: string) {
	const net = getNetwork({
		chainFamily: 'evm',
		chainSelectorName: chainName,
		isTestnet: false,
	});
	if (!net) throw new Error(`Network not found for chain name: ${chainName}`);
	return new cre.capabilities.EVMClient(net.chainSelector.selector);
}

// Safely stringify BigInt
const safeJsonStringify = (obj: unknown) =>
	JSON.stringify(obj, (_, v) => (typeof v === 'bigint' ? v.toString() : v), 2);

// ---------- Reader ----------

function readFeed(
	runtime: Runtime<Config>,
	evmClient: InstanceType<typeof cre.capabilities.EVMClient>,
	name: string,
	address: string,
): PriceResult {
	const aggregator = new PriceFeedAggregator(evmClient, address as Address);
	const decimals = aggregator.decimals(runtime);
	const latestAnswer = aggregator.latestAnswer(runtime);
	const scaled = formatUnits(latestAnswer, decimals);

	runtime.log(
		`Price feed read | chain=${runtime.config.chainName} feed="${name}" address=${address} decimals=${decimals} latestAnswerRaw=${latestAnswer.toString()} latestAnswerScaled=${scaled}`,
	);

	return {
		name,
		address,
		decimals,
		latestAnswerRaw: latestAnswer.toString(),
		scaled,
	};
}

// ---------- Handlers ----------

export function onCron(runtime: Runtime<Config>, _payload: CronPayload): string {
	const evmClient = getEvmClient(runtime.config.chainName);

	const results: PriceResult[] = runtime.config.feeds.map((f) =>
		readFeed(runtime, evmClient, f.name, f.address),
	);

	return safeJsonStringify(results);
}

// ---------- Init ----------

export function initWorkflow(config: Config) {
	const cron = new cre.capabilities.CronCapability();
	return [
		cre.handler(
			cron.trigger({ schedule: config.schedule }),
			onCron,
		),
	];
}
